pkg/load/read.go

```diff
@@ -0,0 +1,221 @@
+	numDigestsToRead  int
+	numReadsPerDigest int
+	concurrency       int
+}
+
+func (self *readAction) String() string {
+	return fmt.Sprintf("read %d:%d:%d", self.numDigestsToRead, self.numReadsPerDigest, self.concurrency)
+}
+
+type readResult struct {
+	startTime time.Time
+	endTime   time.Time
+	success   int
+	errors    int
+}
+
+type readWorkItem struct {
+	actionContext *ActionContext
+	digest        *remote_pb.Digest
+}
+
+type readWorkResult struct {
+	digest *remote_pb.Digest
+	err    error
+}
+
+func readWorker(workChan <-chan *readWorkItem, resultChan chan<- *readWorkResult) {
+	log.Debug("readWorker started")
+	for wi := range workChan {
+		resultChan <- processReadWorkItem(wi)
+	}
+	log.Debug("readWorker stopped")
+}
+
+func processReadWorkItem(wi *readWorkItem) *readWorkResult {
+	// TODO: This should use bytestream if size execeeds the server's batch RPC limit.
+
+	request := remote_pb.BatchReadBlobsRequest{
+		InstanceName: wi.actionContext.InstanceName,
+		Digests: []*remote_pb.Digest{
+			wi.digest,
+		},
+	}
+
+	response, err := wi.actionContext.CasClient.BatchReadBlobs(wi.actionContext.Ctx, &request)
+	if err == nil {
+		s := status.FromProto(response.Responses[0].Status)
+		if s.Code() != codes.OK {
+			return &readWorkResult{
+				digest: wi.digest,
+				err:    s.Err(),
+			}
+		}
+
+		data := response.Responses[0].Data
+		if int64(len(data)) != wi.digest.SizeBytes {
+			return &readWorkResult{
+				digest: wi.digest,
+				err:    fmt.Errorf("size mismatch, expected %d, actual %d", wi.digest.SizeBytes, len(data)),
+			}
+		}
+
+		computedDigest := casutil.ComputeDigest(data)
+		if computedDigest.Hash != wi.digest.Hash {
+			return &readWorkResult{
+				digest: wi.digest,
+				err:    fmt.Errorf("digest hash mismatch"),
+			}
+		}
+	} else {
+		return &readWorkResult{
+			digest: wi.digest,
+			err:    err,
+		}
+	}
+
+	return &readWorkResult{
+		digest: wi.digest,
+		err:    nil,
+	}
+}
+
+func (self *readAction) RunAction(actionContext *ActionContext) error {
+	result := readResult{
+		startTime: time.Now(),
+	}
+
+	workChan := make(chan *readWorkItem)
+	resultChan := make(chan *readWorkResult)
+
+	for c := 0; c < self.concurrency; c++ {
+		go readWorker(workChan, resultChan)
+	}
+
+	knownDigests := actionContext.GetKnownDigests(true)
+
+	var workItems []*readWorkItem
+	for i := 0; i < self.numDigestsToRead; i++ {
+		index := rand.Int() % len(knownDigests)
+		digest := knownDigests[index]
+		for j := 0; j < self.numReadsPerDigest; j++ {
+			workItem := &readWorkItem{
+				actionContext: actionContext,
+				digest:        digest,
+			}
+			workItems = append(workItems, workItem)
+		}
+	}
+
+	// Shuffle the work items.
+	rand.Shuffle(len(workItems), func(i, j int) {
+		tmp := workItems[i]
+		workItems[i] = workItems[j]
+		workItems[j] = tmp
+	})
+
+	// Inject the work items into the channel.
+	go func() {
+		for _, workItem := range workItems {
+			workChan <- workItem
+		}
+
+		close(workChan)
+	}()
```

Comment:
Bear with me (not familiar with go):

Is this spawned because the channel is bounded, and if it isn't spawned, the main goroutine will block? Should the channel not be bounded? That would result in more parallelism, presumably, so maybe that's what you're trying to constrain.

Reply:
It is spawned so that the retrieval of results (which occurs later in this function) is not blocked by injection into the channel used to send work to the worker goroutines. The channel is unbounded but that does not mean it won't block. It is more correct to spawn off the injection and proceed to waiting on the channel where results are written.

Reply:
> The channel is unbounded but that does not mean it won't block.

That is ... surprising!

Reply:
Yeah unbounded channels actually block on each element until something receives the value. https://golang.org/ref/spec#Channel_types

> The capacity, in number of elements, sets the size of the buffer in the channel. If the capacity is zero or absent, the channel is unbuffered and communication succeeds only when both a sender and receiver are ready. Otherwise, the channel is buffered and communication succeeds without blocking if the buffer is not full (sends) or not empty (receives). A nil channel is never ready for communication. 

I should probably switch that code to use a buffered channel with the channel size equal to the number of work items. Then it wouldn't block.

Go ends up being very surprising to me still.